### ISSUE TITLE
Support Pipeline Parallel in Knowledge Distillation

### ISSUE THREAD

**Issue by zhaoyang-star** [NVIDIA/NeMo#11531]
As the [nemo-docs](https://docs.nvidia.com/nemo-framework/user-guide/latest/model-optimization/distillation/distillation.html) says, there are limitations of KD impl in NeMo:

- Only Megatron Core-based GPT models are supported
- Only logit-pair distillation is supported for now
- Pipeline parallelism not yet supported
- FSDP strategy not yet supported

For the 3rd and 4th, both of them are very important features for KD as the teacher model is much larger. TP may cause less training throughput than PP. So will the 2 features be in Roadmap? Thanks. @AAnoosheh @ko3n1g

**Comment by AAnoosheh** [comment]
Hi @zhaoyang-star,

PP will be available soon. First in [Megatron-LM](https://github.com/NVIDIA/Megatron-LM) then followed by NeMo immediately afterwards.

Pytorch FSDP is supported for [Model Optimizer](https://github.com/NVIDIA/TensorRT-Model-Optimizer) based distillation in general, though the Megatron FSDP strategy has not been looked into yet and I cannot provide an ETA right now.

Thank you for the interest!

**Comment by zhaoyang-star** [comment]
@AAnoosheh Thanks for your great work! Recently I used distillation on NeMo. It realy help me a lot. 
Just one more question, how to record `kd_loss` in the tensorboard? `reduced_train_loss` is already recorded on tensorboard. I tried to print `kd_loss` in `/usr/local/lib/python3.10/dist-packages/modelopt/torch/distill/loss_balancers.py` but got no output info.

```python
        aggregate_loss = sum(
            loss * weight for loss, weight in zip(kd_loss_dict.values(), self._kd_loss_weight)
        )
        kd_loss = aggregate_loss
        if output_loss is not None:
            aggregate_loss += (1.0 - sum(self._kd_loss_weight)) * output_loss
        # use 3 ways
        print(f"total_loss: {aggregate_loss}, ce_loss: {output_loss}, kd_loss: {kd_loss}", flush=True)
        print_rank_0(f"total_loss: {aggregate_loss}, ce_loss: {output_loss}, kd_loss: {kd_loss}")
        logging.info(f"total_loss: {aggregate_loss}, ce_loss: {output_loss}, kd_loss: {kd_loss}")

        return aggregate_loss
```

**Comment by AAnoosheh** [comment]
Hi, I need to look into that and see how NeMo's logging works, thanks for letting me know.

**Comment by zhaoyang-star** [comment]
> Hi, I need to look into that and see how NeMo's logging works, thanks for letting me know.

@AAnoosheh Realy thanks for your quick reply. There is few info when google it. Recently I tried to use NeMo Distillation API to train a 8B model and met these 2 problems. The main problem is the trainning time per step is very low, which is ~ 150TFLOPs (TP is enabled) per GPU. While the performance is 450TFLOPs per GPU (PP is enabled) when pretrain it from scratch. I guess it is because NeMo Distillation API only support TP. The machine I used is H800, which the intra bandwidth is half of the H100.

Looking forward to your feedback.

**Comment by SeanLi-OI** [comment]
> Hi [@zhaoyang-star](https://github.com/zhaoyang-star),
> 
> PP will be available soon. First in [Megatron-LM](https://github.com/NVIDIA/Megatron-LM) then followed by NeMo immediately afterwards.
> 
> Pytorch FSDP is supported for [Model Optimizer](https://github.com/NVIDIA/TensorRT-Model-Optimizer) based distillation in general, though the NeMo FSDP strategy has not been looked into yet and I cannot provide an ETA right now.
> 
> Thank you for the interest!

@AAnoosheh Hi, may I inquire if there is a roadmap for Knowledge Distillation in Megatron-LM? I've noticed that there is currently no support for KD in Megatron-LM.

**Comment by AAnoosheh** [comment]
Hi @SeanLi-OI  I just noticed your comment – apologies.

We indeed have KD (with pipeline-parallel support coming soon) in Megatron-LM. Yet not only is it a bit hidden, but the documentation is not where it ideally should be and still mentions the old location of the entrypoint script. We will fix that.

This is [the script](https://github.com/NVIDIA/Megatron-LM/blob/main/examples/export/knowledge_distillation/pretrain_gpt_modelopt.py) and the [documentation for it](https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/inference/docs/distillation.md)

**Comment by AAnoosheh** [comment]
@zhaoyang-star What size is the teacher and student?

**Comment by zhaoyang-star** [comment]
> [@zhaoyang-star](https://github.com/zhaoyang-star) What size is the teacher and student?

The teacher model is 14B and the student model is ~7B. We use H800 80GB SMX.  TP=2 will occure OOM, so TP is set to 4.

**Comment by SeanLi-OI** [comment]
@AAnoosheh Thanks for replies. Looking forward to KD with pipeline-parallel support release.

**Comment by JinXiaozhao** [comment]
What is the default loss_balance function in NeMo?
NeMo already supports distillation training. Does Megatron have plans to support it?
Thanks. @AAnoosheh

**Comment by AAnoosheh** [comment]
NeMo version needs no Loss Balancer by default since it only has one loss (KLDiv on logits)

Megatron-LM supported Distillation even before NeMo, but PP support is still incoming soon for it.
https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/post_training/docs/distillation.md

**Comment by JinXiaozhao** [comment]
@AAnoosheh  Thanks for your reply. Which loss is the reduced_train_loss printed in the NeMo log? lm loss or KD loss？

**Comment by peiyingxin** [comment]
> I forgot to mention, we added support for it last week, but only in NeMo 2.0 [#11766](https://github.com/NVIDIA/NeMo/pull/11766)
> 
> Let me know how it works, we will be optimizing the memory usage in the next couple weeks hopefully as well.

The new feature for pipeline parallelism in knowledge distillation,as demonstrated in [https://github.com/NVIDIA/NeMo/pull/11766](url) ,introduces a new Python script for pipeline parallelism However, this new feature is not compatible with the original tensor parallelism (TP) feature in [https://github.com/NVIDIA/NeMo/blob/main/examples/nlp/language_modeling/megatron_gpt_distillation.py](url) Will pipeline parallelism in knowledge distillation be integrated into [https://github.com/NVIDIA/NeMo/blob/main/examples/nlp/language_modeling/megatron_gpt_distillation.py](url)? I am looking for your reply. @AAnoosheh

**Comment by AAnoosheh** [comment]
@JinXiaozhao it should be only KD loss as LM Loss is not calculated during training steps

@peiyingxin the script you refer to is for NeMo 1.0, while my feature adds KD + PP support for NeMo 2.0  
The NeMo 1.0 script is implicitly deprecated and all new features from our side are developed in NeMo 2.0
